### ISSUE TITLE
Lexer does not display errors yet.

### ISSUE THREAD

**Issue by lukasberglund** [ayazhafiz/hgb#26]
This might be worth discussing, but I think that the compiler should quit when it runs into invalid symbols during lexing. I.e after lexing is done, if there are any invalid symbols, the compiler should display a list of invalid symbols and quit. This would let us parse with the assumption that there are no invalid symbols, which may help with simplicity.

**Comment by ayazhafiz** [comment]
There is another option — parse as far as possible and don’t parse invalid tokens since we know it’s an error anyway. Then we can spit out parse and token errors at once. Either way is good with me; we can go with yours for ease.

**Comment by lukasberglund** [comment]
Yeah let's go with the easy way out for now.

**Comment by ayazhafiz** [comment]
I think we will need a wrapper that runs the lexer, saves invalid token errors, then runs the parser and saves parsing errors, and then returns a parse result + all errors. What do you think?

**Comment by lukasberglund** [comment]
Well right now an issue with that strategy is that the parser will stop when it hits an invalid token, so if our lexer finds an invalid token, the parser is guaranteed to stop early.

**Comment by ayazhafiz** [comment]
We should fix that then...

**Comment by lukasberglund** [comment]
I think fixing that would be really difficult since the parser aborts early on any code that isn't flawless. If we have lexer errors then the code has a flaw. If we skip over invalid symbols we are probably gonna get a parse error since the user probably meant that symbol to mean something. We could have invalid be parsed as wildcards that stand for anything but then we'd have to have some kind of wildcard value in our Grammar. Having such a wildcard would for me be annoying since it isn't elegant but I could accept it.

Alternatively we can abort the lexer on invalid characters like we discussed, which would def lead to more elegant code. We can always add a smarter error strategy later.

**Comment by ayazhafiz** [comment]
I just wonder about the amount of work it would take later to add a smarter error strategy. But you're right, maybe we shouldn't try to solve everything at once. We can stick with the current strategy for now. For a smarter one we may need a different parser anyway. Need to do more research into that.

**Comment by lukasberglund** [comment]
What does "a different parser" mean? Like, can we still keep using Pratt
parsing and the logic we set up for that? Writing the parser I felt like
that was at least 60% of the work.

**Comment by ayazhafiz** [comment]
Yes, we can. It would just have to be modified somewhat to support recovering on missing symbols. You can't see it in the email notification you got for my comment, but I have updated the comment with some links and ideas.

**Comment by lukasberglund** [comment]
Ah, very cool. Yeah I think it would be cool to implement something like RecoveringParserRunner down the line. For now, I say we stick with a ruthless, unforgiving parsing/lexing strategy.

**Comment by ayazhafiz** [comment]
Unforgiving 😂
